### PR TITLE
Vectorize `ARProcess()`

### PR DESCRIPTION
**Pull request by dylanhmorris**
This will allow it to play better with `numpyro.plate` (still not seamlessly, [see here](https://num.pyro.ai/en/stable/primitives.html#scan)), among other things

- This follows the convention that whenever a multi-dim array is a timeseries, time is the first dimension. One reason for that convention (@sbidari maybe worth noting in #443): By default, `numpyro.scan` builds by augmenting the first dimension. So since we most often scan over time, this convention will make `scan`s over time have sensible default output.
- Adds a bunch of tests for correctness
- Adds informative error messages when shapes cannot be broadcast properly
- Adds tests for the errors.
- Note for PRs such as #443: clearest way to take dot products along only certain axes of a multidimensional array of unknown dimensions appears to be via `jnp.einsum`. I think we should use it wherever that's the task https://github.com/CDCgov/PyRenew/pull/439#discussion_r1757223924

### CODE REVIEW

File: pyrenew/process/ar.py [hunk 1]
```diff
@@ -52,60 +48,113 @@ def sample(
...
         Returns
         -------
         ArrayLike
+            of shape (n,) + init_vals.shape[1:].
         """
-        noise_sd_arr = jnp.atleast_1d(noise_sd)
-        if not noise_sd_arr.shape == (1,):
-            raise ValueError("noise_sd must be a scalar. " f"Got {noise_sd}")
         autoreg = jnp.atleast_1d(autoreg)
+        noise_sd = jnp.array(noise_sd)
```
Comment (damonbayer) [inline pyrenew/process/ar.py:62]: Why are these not all `jnp.atleast_1d` or all `jnp.array`?

File: pyrenew/process/ar.py [hunk 2]
```diff
@@ -52,60 +48,113 @@ def sample(
...
+                "to the expected shape of the process "
+                f"history (shape {history_shape}). "
+                "History shape is determined by the "
+                "shapes of the init_vals, autoreg, and "
+                "noise_sd arrays. See ARProcess "
+                "documentation for details"
+            ) from e
+
+        inits_flipped = jnp.flip(inits_broadcast, axis=0)
+        assert jnp.shape(inits_flipped) == history_shape
```
Comment (damonbayer) [inline pyrenew/process/ar.py:110]: Appreciate the informative error messages in this PR, but this one appears to be overlooked.

File: pyrenew/process/ar.py [hunk 3]
```diff
@@ -52,60 +48,113 @@ def sample(
...
+            with numpyro.handlers.reparam(
+                config={noise_name: LocScaleReparam(0)}
+            ):
+                next_noise = numpyro.sample(
+                    noise_name,
+                    numpyro.distributions.Normal(
+                        loc=jnp.zeros(noise_shape), scale=noise_sd
+                    ),
+                )
+                assert jnp.shape(next_noise) == noise_shape
```
Comment (damonbayer) [inline pyrenew/process/ar.py:122]: Better error message?

File: pyrenew/process/ar.py [hunk 4]
```diff
@@ -52,60 +48,113 @@ def sample(
...
+        assert jnp.shape(inits_flipped) == history_shape
+
+        def transition(recent_vals, _):  # numpydoc ignore=GL08
+            with numpyro.handlers.reparam(
+                config={noise_name: LocScaleReparam(0)}
+            ):
+                next_noise = numpyro.sample(
+                    noise_name,
+                    numpyro.distributions.Normal(
+                        loc=jnp.zeros(noise_shape), scale=noise_sd
```
Comment (damonbayer) [inline pyrenew/process/ar.py:119]:
```suggestion
                        scale=noise_sd
```

`loc` is automatically set to `0` of the correct dimension if not provided.

File: pyrenew/process/ar.py [hunk 5]
```diff
@@ -52,60 +48,113 @@ def sample(
...
+                    ),
+                )
+                assert jnp.shape(next_noise) == noise_shape
+
+            dot_prod = jnp.einsum("i...,i...->...", autoreg, recent_vals)
```
Comment (damonbayer) [inline pyrenew/process/ar.py:124]: Haven't seen this before. Should we be looking for other places to use it?

File: pyrenew/process/ar.py [hunk 6]
```diff
@@ -52,60 +48,113 @@ def sample(
...
+
+            new_term = dot_prod + next_noise
+
+            assert new_term.shape == noise_shape
```
Comment (damonbayer) [inline pyrenew/process/ar.py:128]: Better error message? In general I'm wondering when you think these asserts might fail. Can the right shapes be guaranteed by checking only the inputs to our functions, rather than the outputs?

File: pyrenew/process/ar.py [hunk 7]
```diff
@@ -52,60 +48,113 @@ def sample(
...
+                axis=0,
+            )[:order]
+
+            assert new_recent_vals.shape == history_shape
             return new_recent_vals, new_term
 
-        last, ts = scan(transition, init_vals, noise)
-        return jnp.hstack([init_vals, ts])
+        if n > order:
+            last, ts = scan(
```
Comment (damonbayer) [inline pyrenew/process/ar.py:144]: don't need to store `last`.

File: pyrenew/process/ar.py [hunk 8]
```diff
@@ -52,60 +48,113 @@ def sample(
...
+                length=(n - order),
+            )
+
+            ts_with_inits = jnp.concatenate(
+                [inits_broadcast, ts],
+                axis=0,
+            )
+        else:
+            ts_with_inits = inits_broadcast
+        return ts_with_inits[:n]
```
Comment (damonbayer) [inline pyrenew/process/ar.py:157]: light preference for `else: ts = None` and always doing  `ts_with_inits  
 = jnp.concatenate([inits_broadcast, ts], axis=0)`.

File: pyrenew/process/ar.py (hunk 5 above)
Comment (dylanhmorris) [reply pyrenew/process/ar.py:124]: Yes, thanks. Meant to flag this. I think it's the best way to get what we need with dot products but only along certain axes. Will flag in the main PR message.

File: pyrenew/process/ar.py (hunk 4 above)
Comment (dylanhmorris) [reply pyrenew/process/ar.py:119]: It's not currently set anywhere though. Are you suggesting it would be clearer to set via `.expand()` than via the shape of the `loc`?

File: pyrenew/process/ar.py (hunk 1 above)
Comment (damonbayer) [reply pyrenew/process/ar.py:62]: Thinking a bit more.

`autoreg` can be a vector for each `plate` element, while `noise_sd` should be scalar for each `plate` element, so I understand there might be a difference for how they are handled. Not grasping why this is the difference, though.

Comment (dylanhmorris) [reply pyrenew/process/ar.py:62]: `noise_sd` can be a scalar, but the other two need a first dimension (time), even if it's length 1, because that first dimension length gives the order of the process.

Comment (dylanhmorris) [reply pyrenew/process/ar.py:62]: clarifying comment added to code in 3023a3d

File: pyrenew/process/ar.py (hunk 6 above)
Comment (damonbayer) [reply pyrenew/process/ar.py:134]: Understood. I think I prefer to take them out.

File: pyrenew/process/ar.py (hunk 1 above)
Comment (dylanhmorris) [reply pyrenew/process/ar.py]: added a clarifying comment on this point to the code in 3023a3d

File: pyrenew/process/ar.py (hunk 7 above)
Comment (dylanhmorris) [reply pyrenew/process/ar.py]: Addressed in f64c839

File: pyrenew/process/ar.py (hunk 2 above)
Comment (dylanhmorris) [reply pyrenew/process/ar.py]: Addressed in 461e024

File: pyrenew/process/ar.py (hunk 3 above)
Comment (dylanhmorris) [reply pyrenew/process/ar.py]: Addressed in 461e024

File: pyrenew/process/ar.py (hunk 6 above)
Comment (dylanhmorris) [reply pyrenew/process/ar.py]: Done in 461e024

File: pyrenew/process/ar.py (hunk 4 above)
Comment (dylanhmorris) [reply pyrenew/process/ar.py:124]: @damonbayer this is the only question and then should be review for re-review

Comment (damonbayer) [reply pyrenew/process/ar.py:124]: I think I misunderstood initially. I thought `noise_shape` was the shape of `noise_sd`, but it is something else.

**Comment by dylanhmorris** [comment]
> Since one of the intended use cases of this extension is compatibility with `numpyro.plate` I wonder if we should do some tests with `plate`s (and more complicated setups like nested `plate`s. I'm not immediately sure that all this will work).

Good point will add

**Comment by damonbayer** [comment]
> Good point will add

I might even suggest that `plate` be the only way we do multi-dimensional testing.

**Comment by dylanhmorris** [comment]
> > Good point will add
> 
> I might even suggest that `plate` be the only way we do multi-dimensional testing.

I wouldn't go that far; we want vectorizations / multi-dim behavior to be as expected outside of `plate` contexts too, no?

**Review by damonbayer — APPROVED** [review]

**Review by AFg6K7h4fhy2 — COMMENTED** [review]
Confirming that I have seen all changes made in this PR and (mostly) have absorbed the updates.
